pallets/protos/src/lib.rs

```diff
@@ -531,8 +529,65 @@ pub mod pallet {
 	}
 
 	impl<T: Config> Pallet<T> {
-		pub fn get_by_tag(tags: Tags) -> Option<Vec<Hash256>> {
-			<ProtosByTag<T>>::get(&tags)
+
+		fn is_proto_having_any_tags(proto_hash: &Hash256, tags: &Vec<Tags>) -> bool {
+			if let Some(struct_proto) = <Protos<T>>::get(proto_hash) {
+				tags.into_iter().any(|tag| struct_proto.tags.contains(&tag))
+			} else {
+				false
+			}
+		}
+
+		pub fn get_by_tags(tags: Vec<Tags>, owner: Option<T::AccountId>, limit: u32) -> Vec<Hash256> {
+
+
+			// log::info!("inside get_by_tags");
+			// log::info!("tags: {:?}, owner: {:?}, limit: {}", tags, owner, limit);
+
+			match owner {
+				Some(owner) => {
+
+					let iter_protos = <ProtosByOwner<T>>::iter_key_prefix(ProtoOwner::User(owner));
+
+					let iter_protos_filtered = iter_protos.filter(|proto| Self::is_proto_having_any_tags(proto, &tags));
+					let iter_protos_limited = iter_protos_filtered.take(limit as usize);
+					iter_protos_limited.collect::<Vec<Hash256>>()
+
+				},
+				None => {
+					let mut remaining = limit;
+
+					let iter_protos = tags.into_iter().map(|tag| {
+						let vector_protos = <ProtosByTag<T>>::iter_key_prefix(tag).take(remaining as usize).collect::<Vec<Hash256>>();
+						remaining -= vector_protos.len() as u32;
+						vector_protos
+					}).flatten().take(limit as usize);
+
+					iter_protos.collect::<Vec<Hash256>>()
+				}
+
+			}
+
+		}
+
+
+		pub fn get_metadata_batch(batch: Vec<Hash256>, keys: Vec<Vec<u8>>) -> Vec<Option<Vec<Hash256>>> {
+
+			batch.into_iter().map(|proto_hash| -> Option<Vec<Hash256>> {
+
+				let proto = <Protos<T>>::get(&proto_hash)?;
+
+				let mut vec_data_hash : Vec<Hash256> = Vec::new();
+
+				for key in &keys {
+					let data_hash = proto.metadata.get(key)?;
```

Comment:
We need to handle the case where this key is not available actually, instead of `?` let's actually return a `0x0` hash? or we return `Option<Hash256>` instead of `Hash256`.
I'm asking pauline to have placeholders to render.

Reply:
Oh okay. So if it returns:

1. "0x0":  It means the Proto exists but has no metadata
2. None: It means the Proto doesn't exist

Reply:
I think `Option<Hash256>` is cleaner/safer

But then our return type for the function will be `Vec<Option<Vec<Option<Hash256>>>>` !

Reply:
I've changed the return type in the latest commit of this branch. Have a look @sinkingsugar